pkg/ingester/stream.go

```diff
@@ -111,26 +111,39 @@ func (s *stream) consumeChunk(_ context.Context, chunk *logproto.Chunk) error {
 	return nil
 }
 
-func (s *stream) Push(ctx context.Context, entries []logproto.Entry, synchronizePeriod time.Duration, minUtilization float64) error {
+// setChunks is used during checkpoint recovery
+func (s *stream) setChunks(chunks []Chunk) (entriesAdded int, err error) {
+	chks, err := fromWireChunks(s.cfg, chunks)
+	if err != nil {
+		return 0, err
+	}
+	s.chunks = chks
+	for _, c := range s.chunks {
+		entriesAdded += c.chunk.Size()
+	}
+	return entriesAdded, nil
+}
+
+func (s *stream) NewChunk() *chunkenc.MemChunk {
+	return chunkenc.NewMemChunk(s.cfg.parsedEncoding, s.cfg.BlockSize, s.cfg.TargetChunkSize)
+}
+
+func (s *stream) Push(
+	ctx context.Context,
+	entries []logproto.Entry,
+	record *WALRecord,
+) error {
 	var lastChunkTimestamp time.Time
 	if len(s.chunks) == 0 {
 		s.chunks = append(s.chunks, chunkDesc{
-			chunk: s.factory(),
+			chunk: s.NewChunk(),
 		})
 		chunksCreatedTotal.Inc()
 	} else {
 		_, lastChunkTimestamp = s.chunks[len(s.chunks)-1].chunk.Bounds()
 	}
 
-	s.tailerMtx.RLock()
-	hasTailers := len(s.tailers) != 0
-	s.tailerMtx.RUnlock()
-
-	var storedEntries []logproto.Entry
-	if hasTailers {
-		storedEntries = make([]logproto.Entry, 0, len(entries))
-	}
-
+	storedEntries := make([]logproto.Entry, 0, len(entries))
```

Comment:
We're going to have to improve this. this was heavy on allocations

Reply:
Good point. It looks like I can use the pool and release it in the tailer code.